x/twap/store_test.go

```diff
@@ -201,70 +201,177 @@ func (s *TestSuite) TestGetRecordAtOrBeforeTime() {
 }
 
 // TestPruneRecordsBeforeTime tests that all twap records earlier than
-// current block time - given time are pruned from the store.
-func (s *TestSuite) TestPruneRecordsBeforeTime() {
-	tMin2Record, tMin1Record, baseRecord, tPlus1Record := s.createTestRecordsFromTime(baseTime)
-
-	// non-ascending insertion order.
-	allTestRecords := []types.TwapRecord{tPlus1Record, tMin1Record, baseRecord, tMin2Record}
+// current block time - given time are pruned from the store while
+// the newest record for each pool before the time to keep is preserved.
+func (s *TestSuite) TestPruneRecordsBeforeTimeButNewest() {
+	// These are manually created to be able to refer to them by name
+	// for convenience.
+	pool1Min2SBaseMs, pool2Min1SBaseMs, pool3BaseSecBaseMs, pool4Plus1SBaseMs := s.createTestRecordsFromTime(baseTime)
+	pool1Min2SMin1Ms, pool2Min1SMin1Ms, pool3BaseSecMin1Ms, pool4Plus1SMin1Ms := s.createTestRecordsFromTime(baseTime.Add(-time.Millisecond))
+	pool1Min2SMin2Ms, pool2Min1SMin2Ms, pool3BaseSecMin2Ms, pool4Plus1SMin2Ms := s.createTestRecordsFromTime(baseTime.Add(2 * -time.Millisecond))
+	pool1Min2SMin3Ms, pool2Min1SMin3Ms, pool3BaseSecMin3Ms, pool4Plus1SMin3Ms := s.createTestRecordsFromTime(baseTime.Add(3 * -time.Millisecond))
+
+	// Create records that match times 1:1 with other pools but have a different pool ID.
+	pool5Min2SBaseMs, pool5Min1SBaseMs, pool5BaseSecBaseMs, pool5Plus1SBaseMs := s.createTestRecordsFromTime(baseTime)
```

Comment:
im wondering if there's reason we need to test with so many pools, since I think max 3 pools would suffice testing any case possible with multiple pools

Reply:
They all start at the different base times:
https://github.com/osmosis-labs/osmosis/blob/2c7649f4ed193e05ae921c65cb2a037d9c3412d3/x/twap/store_test.go#L209-L212


The goal is not only to test the inter-pool correctness but also how pruning works with respect to different times